File: homework3/Aditya_Prasad/q7_ReverseWords.cpp
// DS/Algo used - Stack

// Time Complexity O(n)

// Memory Complexity O(n)

// where n is the size of given string 

#include <bits/stdc++.h>
using namespace std;

string ReverseWords(string words){
    stack<string> st;
    string word;

    for (char c : words){
        if (c == ' '){
            st.push(word);
            word = "";
        }else{
            word += c;
        }
    }

    st.push(word);

    string ReversedString;

    while (!st.empty()){
        ReversedString += st.top();
        st.pop();
        if(!st.empty()) ReversedString += " ";
    }

    return ReversedString;
}

int main(){
    cout << ReverseWords("Uber Career Prep") << endl;
    cout << ReverseWords("Emma lives in Brooklyn, New York.");
}

// Time Taken - 9 min

File: homework3/Satyam_Mishra/Q7_ReverseWords.java
// Assumptions
// 1. only single space is in between

// DS/Algo used - TRAVERSAL

// Time Complexity = O(N), where R is length of road array
// Space Complexity = O(N)

package homework3.Satyam_Mishra;

public class Q7_ReverseWords {
    private static String reverseWords(String words) {

        String[] wordsArray = words.split(" ");

        StringBuilder sb = new StringBuilder();
        for(int i = wordsArray.length - 1; i >= 0; i--) {
            sb.append(wordsArray[i]);
            if (i > 0) {
                sb.append(" ");
            }
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        String words1 = "Uber Career Prep";
        System.out.println("Reversed words1 is: " + reverseWords(words1));

        String words2 = "Emma lives in Brooklyn, New York";
        System.out.println("Reversed words2 is: " + reverseWords(words2));
    }
}

// time taken 8m 20s